### PR TITLE
sql: repurpose the syntax x'...' for byte array literals.

### LINKED ISSUE #13049: sql: syntax X'...' should be byte array not string
CockroachDB supports the non-standard extension syntax `x'.....'` to enter a literal string using hexadecimal codes only.

This is all fine and dandy but it's not so clear to me that we have a use case for it.

Meanwhile, we do have a use case for *entering raw bytes containing possibly invalid UTF-8 sequences (a byte array) in hexadecimal*, for example a uuid value or a SHA hash, but CockroachDB does not provide a compact syntax to do so yet -- the best we have so far is `b'\xAA\xAA\xAA\xAA...` which is 50% less efficient to type.

It seems to me we should simply repurpose the syntax `x'....'` to become a byte array literal (BCONST) instead of a string array literal (SCONST). 

@jordanlewis what do you think? (You added this support initially)

cc @nvanbenschoten

### PR DESCRIPTION
**Pull request by knz**
Prior to this patch this syntax would produce string literals.  Like
for all string values it would also check that the string does not
contain invalid UTF-8 sequences, so it was not usable to enter byte
array constants. Meanwhile there was no convenient syntax to enter
arbitrary byte sequences using hex codes only (like hash values,
binary encoded data, etc).

As this feature was originally intended to mimic MySQL's similar
syntax, which does serve to produce byte arrays, this patch aligns the
feature more closely with its original intent by ensuring the syntax
produces byte array literals instead of string literals.

Its usage for string literals is still possible however by means of a
simple cast.

Fixes #13049.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13094)

### CODE REVIEW

**Comment by jordanlewis** [comment]
Like magic! LGTM. Could you add a test case illustrating the cast to string, just for completeness?

**Comment by knz** [comment]
> Like magic! LGTM. Could you add a test case illustrating the cast to
> string, just for completeness?
> 

Done

**Comment by knz** [comment]
TFYR :)

File: pkg/sql/parser/parse_test.go [hunk 1]
```diff
@@ -1042,7 +1042,7 @@ SELECT 0x FROM t
 		},
 		{
 			`SELECT x'fail' FROM t`,
```
Comment (mjibson) [inline pkg/sql/parser/parse_test.go]: I'd also like to see an error case with the valid characters but wrong number of characters: `SELECT x'01a'`

Comment (knz) [inline pkg/sql/parser/parse_test.go]: Done.

**Pull request merged**